geth/peers/topicpool.go

```diff
@@ -112,40 +171,55 @@ func (t *TopicPool) ConfirmAdded(server *p2p.Server, nodeID discover.NodeID) {
 func (t *TopicPool) ConfirmDropped(server *p2p.Server, nodeID discover.NodeID) bool {
 	t.mu.Lock()
 	defer t.mu.Unlock()
+
+	discV5NodeID := discv5.NodeID(nodeID)
+
 	// either inbound or connected from another topic
-	peer, exist := t.peers[discv5.NodeID(nodeID)]
+	peer, exist := t.connectedPeers[discV5NodeID]
 	if !exist {
 		return false
 	}
-	log.Debug("disconnect", "ID", nodeID)
-	if peer.requested {
-		return false
-	}
-	if t.SearchRunning() && t.connected == t.limits[0] {
+
+	log.Debug("disconnect", "ID", nodeID, "dismissed", peer.dismissed)
+
+	// switch to fast mode as the number of connected peers is about to drop
+	// below the lower limit
+	if t.SearchRunning() && len(t.connectedPeers) == t.limits[0] {
 		t.period <- t.fastSync
 	}
-	t.connected--
-	t.removePeer(server, peer)
```

Comment:
As `ConfirmDropped` is called when a peer was removed from the server, this line looks obsolete.

Reply:
> t.removePeer(server, peer)
needed because we still need to remove a peer from static set, event will be received when connection is lost

> connected--

how then connected count will be decremented?

Reply:
You mean it's not removed from the node's server static peers set?

Reply:
yes, for example, we can receive event that there was a network error or we were intentionally disconnected, but because AddPeer also adds a peer to static set - peer will stay in server static set and p2p server will try to reconnect with it after 30s(?).